### ISSUE TITLE
augment does not always return `.rownames` column

### ISSUE THREAD

**Issue by mpettis** [tidymodels/broom#168]
Function `augment` does not always return a `.rownames` column.   It seems to happen only when not all of the rows in the source dataframe for a linear model can be used in the linear model (due to missing values).  Is there a way to get an option in `augment` to tell it to always return the `.rownames` as a column?  Also, is this behavior documented somewhere?  My scan of the documentation did not reveal this to me.  Otherwise, I have been relying on that column being present.  Below is an example showing the phenomenon.

`

> dtmp <- data_frame(x=seq(0, 1, 0.1), y=2 \* x + rnorm(length(x), 0, 0.1))
> broom::augment(lm(y ~ x, dtmp)) %>% names()
> [1] "y"          "x"          ".fitted"    ".se.fit"    ".resid"     ".hat"       ".sigma"     ".cooksd"    ".std.resid"
> 
> dtmp <- data_frame(x=seq(0, 1, 0.1), y=2 \* x + rnorm(length(x), 0, 0.1))
> dtmp[5, 2] <- NA
> broom::augment(lm(y ~ x, dtmp)) %>% names()
>  [1] ".rownames"  "y"          "x"          ".fitted"    ".se.fit"    ".resid"     ".hat"       ".sigma"     ".cooksd"  
> [10] ".std.resid"
> `

**Comment by dgrtwo** [comment]
This behavior is intentional, though I can see how it's confusing. The .rownames column is ignored if the data frame doesn't have rownames, which is determined based on whether the rownames are equivalent to seq_len(nrow(d)). Otherwise, many data frames without rownames would end up with a pretty redundant .rownames column.

This does lead to the slightly odd behavior when NAs are included, since lm turns it into a regular data frame before doing the subsetting, which puts the rownames back and makes them non-sequential.

Generally with tbl_dfs, it's good to avoid relying on rownames at all (better to add a column beforehand- I like doing `mutate(observation = row_number())` that you can use to match them together).

I'll leave this open to remind myself to add better documentation about when .rownames is added.

**Comment by mpettis** [comment]
Thanks.  And yes, I'd like to reliably merge the output of `augment` back onto the original data frame that was used by lm (which may have more rows than were actually used in the fit).  So, to be clear, if there is no `.rownumber` on the `augment` output, can I use `row_number()` on the `augment` dataframe output as a key to merge back onto the source dataframe to the `lm` function?  I just want to make sure I can reliably match the `augment` output back to the source.

Thanks,
Matt

**Comment by alexpghayes** [comment]
```r
library(dplyr)
library(broom)

df <- data.frame(x = 1:10, y = 2 * (1:10) + 5 + rnorm(10))

df <- df %>% 
  mutate(id = row_number())  # add a unique row identifier

df_na <- df
df_na[5, 2] <- NA

# be sure to pass `df` to the data argument to get the id column
# can now match on id column
lm(y ~ x, df) %>% 
  augment(data = df)  

# same approach works if there is missing data, although now row 5 disappears
lm(y ~ x, df_na) %>% 
  augment(data = df_na)  
```

But we're going to change the behavior of `augment` so that it always returns *all* of the observations in data, and then this won't be an issue anymore.

TODO: still need to document the `.rownames` behavior.

**Comment by IndrajeetPatil** [comment]
@alexpghayes The second example from your reprex no longer works, so maybe that already addresses this?

``` r
library(dplyr)
library(broom)

df <- data.frame(x = 1:10, y = 2 * (1:10) + 5 + rnorm(10))

df <- df %>% 
  mutate(id = row_number())  # add a unique row identifier

df_na <- df
df_na[5, 2] <- NA

# be sure to pass `df` to the data argument to get the id column
# can now match on id column
lm(y ~ x, df) %>% 
  augment(data = df)  
#> # A tibble: 10 x 9
#>        x     y    id .fitted  .resid .std.resid  .hat .sigma .cooksd
#>    <int> <dbl> <int>   <dbl>   <dbl>      <dbl> <dbl>  <dbl>   <dbl>
#>  1     1  6.89     1    7.07 -0.177      -0.413 0.345  0.561 0.0449 
#>  2     2  9.07     2    8.99  0.0778      0.169 0.248  0.566 0.00474
#>  3     3 11.5      3   10.9   0.577       1.20  0.176  0.514 0.153  
#>  4     4 12.5      4   12.8  -0.367      -0.741 0.127  0.547 0.0401 
#>  5     5 14.6      5   14.8  -0.195      -0.388 0.103  0.562 0.00864
#>  6     6 16.0      6   16.7  -0.654      -1.30  0.103  0.503 0.0973 
#>  7     7 19.6      7   18.6   0.968       1.95  0.127  0.410 0.279  
#>  8     8 20.7      8   20.5   0.181       0.376 0.176  0.562 0.0151 
#>  9     9 21.9      9   22.4  -0.541      -1.18  0.248  0.516 0.229  
#> 10    10 24.5     10   24.4   0.130       0.303 0.345  0.564 0.0242

# same approach works if there is missing data, although now row 5 disappears
lm(y ~ x, df_na) %>% 
  augment(data = df_na)  
#> Error: Assigned data `predict(x, na.action = na.pass, ...) %>% unname()` must be compatible with existing data.
#> x Existing data has 10 rows.
#> x Assigned data has 9 rows.
#> i Only vectors of size 1 are recycled.
```

<sup>Created on 2020-06-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by savadhanistablab** [comment]
Hi, I had a similar issue here, but for some reason, the augment(fit) (where the fit is say a sample `lm` model) doesn't  include the .se.fit column! And I've tried reinstalling the broom package a couple of times now, yet it doesn't seem to work as expected. I also noticed that similar is the case in the above example by @IndrajeetPatil. 

Incidentally, the package documentation of the broom package too doesn't mention any changes around having the `.se.fit` in the `augment` function.
